### PR TITLE
Adds Code Coverage

### PR DESCRIPTION
**Pull request by navinns**
*Issue #, if available:*

*Description of changes:*
* Adds code to create coverage report
* Adds option to upload report to codecov.io

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tools/ci/ci_agent.py [hunk 1]
```diff
@@ -88,6 +88,17 @@ def cli_run(args):
     _check_status(result)
 
 
+def cli_code_coverage(args):
```
Comment (aggarw13) [inline tools/ci/ci_agent.py:91]: Any (one-line) comment(s) around functions?

Comment (navinns) [reply tools/ci/ci_agent.py:91]: I will update all docs together

File: tools/ci/ci_agent.py [hunk 2]
```diff
@@ -278,6 +310,33 @@ def _run_targets(targets, src, build_path, config):
...
+        target_result = result.setdefault("coverage", {})
+        target_build_result = target_result.setdefault("CodeCoverage", {})
+        out = _build_target("coverage", src, build_path, build_flags, c_flags)
+        log(out)
+
+        if codecov_token:
+            log("\n----------------------------------------------------------------")
+            log("Upload Code Coverage report to codecov.io")
+            log("----------------------------------------------------------------")
+            cmd = f"curl -s https://codecov.io/bash | bash -s - -t {codecov_token}"
```
Comment (dan4thewin) [inline tools/ci/ci_agent.py:325]: Hey... umm...  this is a non-starter.

Comment (dan4thewin) [reply tools/ci/ci_agent.py:325]: To clarify, I mean running untrusted code is a non-starter.

File: tools/ci/ci_agent.py [hunk 3]
```diff
@@ -278,6 +312,68 @@ def _run_targets(targets, src, build_path, config):
...
 
 
+def _build_code_coverage(src, build_path, build_flags, c_flags, codecov_token):
+    result = {}
+    try:
+        target_result = result.setdefault("coverage", {})
+        target_build_result = target_result.setdefault("CodeCoverage", {})
+        out = _build_target("coverage", src, build_path, build_flags, c_flags)
+        log(out)
+        out = _run_cmd(f"gcovr -r . -x -o {build_path}/cobertura.xml")
```
Comment (nateglims) [inline tools/ci/ci_agent.py:322]: Does this need to be named cobertura?

File: tools/ci/ci_agent.py [hunk 4]
```diff
@@ -278,6 +312,68 @@ def _run_targets(targets, src, build_path, config):
...
+        log(out)
+
+        commit_id = os.environ.get("ghprbActualCommit") or ""
```
Comment (nateglims) [inline tools/ci/ci_agent.py:325]: We probably want coverage run on the merge head (`${sha1}` in Jenkins)

File: tools/ci/ci_agent.py [hunk 5]
```diff
@@ -159,6 +180,19 @@ def add_arguments(cmd, *args):
...
+    cmd_code_coverage = subparsers.add_parser("code-coverage")
+    add_arguments(
+        cmd_code_coverage,
+        "--config-file",
+        "--src",
+        "--build-path",
+        "--build-flags",
+        "--c-flags",
+        "--codecov-token",
+    )
```
Comment (nateglims) [inline tools/ci/ci_agent.py:193]: At this point I'm inclined to think we should just take the config as an argument and do the rest through the config or environment variables. It's getting difficult to grapple with all the options.

File: tools/ci/ci_agent.py [hunk 6]
```diff
@@ -278,6 +312,68 @@ def _run_targets(targets, src, build_path, config):
...
+        else:
+            if not commit_id:
+                log(
+                    "Please provide commit id, if you want to upload coverage report to codecov.io"
+                )
+            if not codecov_token:
+                log(
+                    "Provide '--codecov-token' in commandline, \
+                    if you want to upload coverage report to codecov.io"
+                )
```
Comment (nateglims) [inline tools/ci/ci_agent.py:368]: These appear to also be in either the environment or the config file?

File: tools/ci/ci_agent.py (hunk 1 above)
Comment (nateglims) [reply tools/ci/ci_agent.py:93]: We should do doc strings at PR time. They should describe the arguments also. Especially when passing a dict like this.

File: tools/ci/ci_agent.py [hunk 7]
```diff
@@ -278,6 +312,68 @@ def _run_targets(targets, src, build_path, config):
     return result
 
 
+def _build_code_coverage(src, build_path, build_flags, c_flags, codecov_token):
```
Comment (nateglims) [inline tools/ci/ci_agent.py:315]: what's the significance of `build_flags` vs `c_flags`?

File: tools/ci/ci_agent.py (hunk 5 above)
Comment (navinns) [reply tools/ci/ci_agent.py:193]: config is one of the argument, we have options to override options in config

Comment (nateglims) [reply tools/ci/ci_agent.py:193]: I'm not convinced that flexibility is required.

File: tools/ci/ci_agent.py (hunk 6 above)
Comment (navinns) [reply tools/ci/ci_agent.py:368]: yes it is present in config file

File: tools/ci/ci_agent.py (hunk 7 above)
Comment (navinns) [reply tools/ci/ci_agent.py:315]: in csdk we have 2 levels of flags one for cmake and other for compiler

build_flags is for cmake and c_flags are for compiler

File: tools/ci/ci_agent.py (hunk 3 above)
Comment (navinns) [reply tools/ci/ci_agent.py:322]: yes as we can use this file with cobertura as well

File: tools/ci/ci_agent.py (hunk 4 above)
Comment (navinns) [reply tools/ci/ci_agent.py:325]: hmm ok

File: tools/ci/ci_agent.py [hunk 8]
```diff
@@ -115,6 +133,9 @@ def add_arguments(cmd, *args):
         "--c-flags", nargs="+", help="Optional c_flags required to configure build",
     )
     new_argument("--allow", nargs="+", help="Pattern for target selection")
+    new_argument(
+        "--codecov-token", help="Optional token to uplode coverage report to codecov.io"
```
Comment (pavanmr94) [inline tools/ci/ci_agent.py:137]: `uplode` -> `upload`

File: tools/ci/ci_agent.py (hunk 7 above)
Comment (nateglims) [reply tools/ci/ci_agent.py:315]: I don't like the build_flags name. But at the very least we should have a descriptive doc string saying what they are.

File: tools/ci/ci_agent.py (hunk 8 above)
Comment (navinns) [reply tools/ci/ci_agent.py:137]: done

File: tools/ci/ci_agent.py (hunk 1 above)
Comment (nateglims) [reply tools/ci/ci_agent.py:93]: Looks good.

File: tools/ci/ci_agent.py [hunk 9]
```diff
@@ -278,6 +326,78 @@ def _run_targets(targets, src, build_path, config):
...
+    c_flags: Array of c flags used by compiler
+    codecov_token: codecov token required to upload code to codecov.io
+    """
+    result = {}
+    try:
+        target_result = result.setdefault("coverage", {})
+        target_build_result = target_result.setdefault("CodeCoverage", {})
+        out = _build_target("coverage", src, build_path, build_flags, c_flags)
+        log(out)
+        out = _run_cmd(f"gcovr -r . -x -o {build_path}/cobertura.xml")
```
Comment (dan4thewin) [inline tools/ci/ci_agent.py:346]: Looks like this might be passed to a shell - so what happens if build path contains a space?

Comment (dan4thewin) [reply tools/ci/ci_agent.py:346]: or worse, a semicolon...

**Pull request merged**